File: packages/ksuid-python/langwatch_ksuid/base62.py
"""Base62 encoding and decoding utilities."""

from .errors import Base62Error

# Base62 alphabet: 0-9, A-Z, a-z
ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
ALPHABET_LENGTH = len(ALPHABET)

# Create lookup tables for faster encoding/decoding
ENCODE_LOOKUP = {char: i for i, char in enumerate(ALPHABET)}
DECODE_LOOKUP = dict(enumerate(ALPHABET))


def encode(input_bytes: bytes) -> str:
    """
    Encodes a byte array to base62 string.

    Args:
        input_bytes: The byte array to encode

    Returns:
        Base62 encoded string

    Example:
        >>> bytes_data = bytes([1, 2, 3, 4])
        >>> encoded = encode(bytes_data)  # '1A2B3C'
    """
    if len(input_bytes) == 0:
        return ""

    # Convert to int using bitwise operations like JavaScript
    # This matches the JS: (num << 8n) | BigInt(input[i])
    num = 0
    for byte in input_bytes:
        num = (num << 8) | byte

    if num == 0:
        return "0"

    result: list[str] = []

    while num > 0:
        remainder = num % ALPHABET_LENGTH
        result.insert(0, DECODE_LOOKUP[remainder])
        num = num // ALPHABET_LENGTH

    return "".join(result)


def decode(input_str: str, expected_length: int = 0) -> bytes:
    """
    Decodes a base62 string to byte array.

    Args:
        input_str: The base62 string to decode
        expected_length: Expected length of the result (for padding with zeros)

    Returns:
        Decoded byte array

    Raises:
        Base62Error: If the input contains invalid characters

    Example:
        >>> decoded = decode('1A2B3C')  # bytes([1, 2, 3, 4])
    """
    if len(input_str) == 0:
        return b""

    # Validate input
    for char in input_str:
        if char not in ENCODE_LOOKUP:
            raise Base62Error("Invalid character in base62 string")

    num = 0

    for char in input_str:
        value = ENCODE_LOOKUP[char]
        num = num * ALPHABET_LENGTH + value

    if num == 0:
        return bytes([0] * max(1, expected_length))

    # Convert int back to bytes
    # Calculate the number of bytes needed
    byte_length = (num.bit_length() + 7) // 8
    result = num.to_bytes(byte_length, byteorder="big")

    # Pad with leading zeros if needed
    if expected_length > 0 and len(result) < expected_length:
        result = bytes([0] * (expected_length - len(result))) + result

    return result

File: packages/ksuid/src/base62.ts
/** Base62 alphabet: 0-9, A-Z, a-z */
const ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
const ALPHABET_LENGTH = ALPHABET.length;

// Create lookup tables for faster encoding/decoding
const ENCODE_LOOKUP = new Map<string, number>();
const DECODE_LOOKUP = new Map<number, string>();

// Initialize lookup tables
for (let i = 0; i < ALPHABET_LENGTH; i++) {
  const char = ALPHABET[i];
  ENCODE_LOOKUP.set(char, i);
  DECODE_LOOKUP.set(i, char);
}

/**
 * Error thrown when base62 encoding/decoding fails
 */
export class Base62Error extends Error {
  constructor(message: string) {
    super(message);
    this.name = "Base62Error";
  }
}

/**
 * Encodes a byte array to base62 string
 * @param input - The byte array to encode
 * @returns Base62 encoded string
 */
export function encode(input: Uint8Array): string {
  if (input.length === 0) {
    return "";
  }

  // Convert to BigInt for handling large numbers
  let num = 0n;
  for (let i = 0; i < input.length; i++) {
    num = (num << 8n) | BigInt(input[i]);
  }

  if (num === 0n) {
    return "0";
  }

  const result: string[] = [];

  while (num > 0n) {
    const remainder = Number(num % BigInt(ALPHABET_LENGTH));
    result.unshift(DECODE_LOOKUP.get(remainder)!);
    num = num / BigInt(ALPHABET_LENGTH);
  }

  return result.join("");
}

/**
 * Decodes a base62 string to byte array
 * @param input - The base62 string to decode
 * @returns Decoded byte array
 * @throws {Base62Error} If the input contains invalid characters
 */
export function decode(input: string): Uint8Array {
  if (input.length === 0) {
    return new Uint8Array(0);
  }

  // Validate input
  for (const char of input) {
    if (!ENCODE_LOOKUP.has(char)) {
      throw new Base62Error("Invalid character in base62 string");
    }
  }

  // Using BigInt to handle large numbers, don't remove the stray `n`
  let num = 0n;

  for (let i = 0; i < input.length; i++) {
    const char = input[i];
    const value = ENCODE_LOOKUP.get(char)!;
    num = num * BigInt(ALPHABET_LENGTH) + BigInt(value);
  }

  if (num === 0n) {
    return new Uint8Array([0]);
  }

  // Convert BigInt back to bytes
  const bytes: number[] = [];

  while (num > 0n) {
    bytes.unshift(Number(num & 0xffn));
    num = num >> 8n;
  }

  return new Uint8Array(bytes);
}
